rclcpp/src/rclcpp/executors/static_single_threaded_executor.cpp

```diff
@@ -0,0 +1,329 @@
+          exec_list.service.push_back(service);
+          exec_list.number_of_services++;
+        }
+        return false;
+      });
+    }
+  }
+}
+
+void
+StaticSingleThreadedExecutor::get_client_list(ExecutableList & exec_list)
+{
+  // Clear the previous clients (if any) from the ExecutableList struct
+  exec_list.client.clear();
+  exec_list.number_of_clients = 0;
+  for (auto & weak_node : weak_nodes_) {
+    auto node = weak_node.lock();
+    if (!node) {
+      continue;
+    }
+    // Check in all the callback groups
+    for (auto & weak_group : node->get_callback_groups()) {
+      auto group = weak_group.lock();
+      if (!group || !group->can_be_taken_from().load()) {
+        continue;
+      }
+      group->find_client_ptrs_if([&exec_list](const rclcpp::ClientBase::SharedPtr & client) {
+        if(client){
+          //if any client is found, push it in the exec_list struct
+          exec_list.client.push_back(client);
+          exec_list.number_of_clients++;
+        }
+        return false;
+      });
+    }
+  }
+}
+
+void
+StaticSingleThreadedExecutor::get_waitable_list(ExecutableList & exec_list)
+{
+  // Clear the previous waitables (if any) from the ExecutableList struct
+  exec_list.waitable.clear();
+  exec_list.number_of_waitables = 0;
+  for (auto & weak_node : weak_nodes_) {
+    auto node = weak_node.lock();
+    if (!node) {
+      continue;
+    }
+    // Check in all the callback groups
+    for (auto & weak_group : node->get_callback_groups()) {
+      auto group = weak_group.lock();
+      if (!group || !group->can_be_taken_from().load()) {
+        continue;
+      }
+      group->find_waitable_ptrs_if([&exec_list](const rclcpp::Waitable::SharedPtr & waitable) {
+        if(waitable){
+          //if any waitable is found, push it in the exec_list struct
+          exec_list.waitable.push_back(waitable);
+          exec_list.number_of_waitables++;
+        }
+          return false;
+      });
+    }
+  }
+}
+
+void
+StaticSingleThreadedExecutor::get_executable_list(
+  ExecutableList & executable_list, std::chrono::nanoseconds timeout)
+{
+  // prepare the wait_set
+  prepare_wait_set();
+
+  // add handles to the wait_set and wait for work
+  refresh_wait_set(timeout);
+
+  // Get all the timers
+  get_timer_list(executable_list);
+
+  // Get all the subscribers
+  get_subscription_list(executable_list);
+
+  // Get all the services
+  get_service_list(executable_list);
+
+  // Get all the clients
+  get_client_list(executable_list);
+
+  // Get all the waitables
+  get_waitable_list(executable_list);
+}
+
+void
+StaticSingleThreadedExecutor::execute_ready_executables(
+  ExecutableList & exec_list, std::chrono::nanoseconds timeout)
+{
+    refresh_wait_set(timeout);
+    //Execute all the ready subscriptions
+    for (size_t i = 0; i < wait_set_.size_of_subscriptions; ++i) {
+      if (wait_set_.size_of_subscriptions && i < exec_list.number_of_subscriptions) {
```

Comment:
check exec_list.number_of_subscriptions at 1st, so we do not do that every single loop of wait_set_.size_of_subscriptions

Reply:
> check exec_list.number_of_subscriptions at 1st, so we do not do that every single loop of wait_set_.size_of_subscriptions ? also this applies to time, service and clients.

@fujitatomoya thanks for the suggestion, we would consider making this change.